crates/wasmtime/src/runtime/vm/gc/enabled/free_list.rs

```diff
@@ -37,8 +36,62 @@ impl FreeList {
         free_list
     }
 
+    /// Add additional capacity to this free list.
+    #[allow(dead_code)] // TODO: becomes used in https://github.com/bytecodealliance/wasmtime/pull/10503
+    pub fn add_capacity(&mut self, additional: usize) {
+        let old_cap = self.capacity;
+        self.capacity = self.capacity.saturating_add(additional);
+        log::debug!(
+            "FreeList::add_capacity({additional:#x}): capacity growing from {old_cap:#x} to {:#x}",
+            self.capacity
+        );
+
+        // If we are adding capacity beyond what a `u32` can address, then we
+        // can't actually use that capacity, so don't bother adding a new block
+        // to the free list.
+        let old_cap_rounded = round_usize_down_to_pow2(old_cap, ALIGN_USIZE);
```

Comment:
If allocations could previously go up to `old_cap`, shouldn't this round up instead of round down?

Reply:
No, because the old capacity was also rounded down the the alignment (all allocations are also a multiple of the alignment) so the first newly allocatable index for our additional capacity is that old capacity rounded down.

Reply:
Should this assert that old_cap is already aligned in that case? It mostly just seems counter-inutitive to me that you'd round down the end to figure out the index of the next block, I'd expect that you'd either round up or assert rounding isn't necessary

Reply:
If we keep `self.capacity` aligned down to our alignment, then a sequence of 100 `add_capacity(1)` calls will never actually result in the `FreeList` thinking it has capacity for an allocation, even though it actually should. That is, `self.capacity = round_down_to_align(self.capacity + additional)` will always result in `self.capacity ==0` when `self.capacity` is initially `0` and `additional` is `1`, even when it is called in a loop.

We have to keep track of the actual capacity, and just do the rounding when adding blocks into the free list.

Reply:
Ah ok yeah that makes sense, I was missing that subtelty. Mind adding some comments to that effect to this?

Reply:
Added